### PR TITLE
CY-2184 - System status icon not up to date

### PR DESCRIPTION
**Pull request by qooban**
Implemented approach:
* cluster status is fetched periodically, every 10 seconds, response from REST API is saved in redux store, status fetching is optimized this way:
  * when Cluster Status widget is on the active page, full cluster status is fetched
  * when there is no Cluster Status widget on the active page, only summary cluster status is fetched
* Cluster Status widget is not polling data at all, it is reusing data from redux store

### CODE REVIEW

File: app/actions/clusterStatus.js [hunk 1]
```diff
@@ -22,15 +22,28 @@ export function errorClusterStatus(error) {
     };
 }
 
-export function getClusterStatus() {
+function isClusterStatusWidgetOnPage(pageId, pages) {
+    const currentPage = _.find(pages, page => page.id === pageId);
```
Comment (kubama) [inline app/actions/clusterStatus.js]: If `pageId` is string or number it could be simplified to `const currentPage = _.find(pages, {id: pageId});`

File: app/actions/clusterStatus.js [hunk 2]
```diff
@@ -22,15 +22,28 @@ export function errorClusterStatus(error) {
...
+    const clusterStatusWidgetDefinitionName = 'highAvailability';
+    const clusterStatusWidgets = _.find(
+        _.get(currentPage, 'widgets'),
+        widget => widget.definition === clusterStatusWidgetDefinitionName
```
Comment (kubama) [inline app/actions/clusterStatus.js]: Assuming `clusterStatusWidgetDefinitionName` is a string line 30 could be simplified to:
`{definition: clusterStatusWidgetDefinitionName}`

File: app/actions/clusterStatus.js [hunk 3]
```diff
@@ -22,15 +22,28 @@ export function errorClusterStatus(error) {
...
+    );
+
+    return !_.isUndefined(clusterStatusWidgets);
```
Comment (kubama) [inline app/actions/clusterStatus.js]: To omit `!` operator I would change this line to `return _.isObject(clusterStatusWidgets);`

File: app/reducers/clusterStatusReducer.js [hunk 4]
```diff
@@ -9,7 +9,7 @@ const clusterStatus = (state = {}, action) => {
                 isFetching: false,
                 error: undefined,
                 status: action.status,
-                services: action.services
+                services: action.services !== undefined ? action.services : state.services
```
Comment (kubama) [inline app/reducers/clusterStatusReducer.js]: Could it be simplified to `services: action.services || state.services`?

File: app/utils/StatusPoller.js [hunk 5]
```diff
@@ -12,6 +13,7 @@ export default class StatusPoller {
         this._store = store;
         this._pollerTimer = null;
         this._fetchStatusPromise = null;
+        this._fetchClusterStatusPromise = null;
```
Comment (kubama) [inline app/utils/StatusPoller.js]: This violates `no-underscore-dangle` eslint rule. This currently gives a warning, but is listed in `.eslintrc` as a candidate for error once all occurrences are fixed. If so I think we shouldn't introduce more violations of this rule.

File: package.json [hunk 6]
```diff
@@ -82,14 +82,14 @@
...
-    "@babel/register": "^7.5.5",
-    "@babel/runtime": "^7.5.5",
+    "@babel/core": "^7.7.7",
+    "@babel/plugin-proposal-class-properties": "^7.7.4",
+    "@babel/plugin-transform-runtime": "^7.7.6",
+    "@babel/polyfill": "^7.7.0",
+    "@babel/preset-env": "^7.7.7",
+    "@babel/preset-react": "^7.7.4",
+    "@babel/register": "^7.7.7",
+    "@babel/runtime": "^7.7.7",
```
Comment (kubama) [inline package.json]: Are these upgrades related to the code changes? If not I don't think it is a good idea to make such upgrades in a pre-release patch. I'm aware these are dev dependencies but as I understand they are used in project build and as such influence the whole application providing some level of risk.

File: app/actions/clusterStatus.js [hunk 7]
```diff
@@ -22,15 +22,28 @@ export function errorClusterStatus(error) {
     };
 }
 
-export function getClusterStatus() {
+function isClusterStatusWidgetOnPage(pageId, pages) {
+    const currentPage = _.find(pages, page => page.id === pageId);
+    const clusterStatusWidgetDefinitionName = 'highAvailability';
+    const clusterStatusWidgets = _.find(
```
Comment (kubama) [inline app/actions/clusterStatus.js]: Since `_.find` returns a single object I would rather use singular form for this const: `clusterStatusWidget`

File: app/actions/clusterStatus.js [hunk 8]
```diff
@@ -22,15 +22,28 @@ export function errorClusterStatus(error) {
...
+export function getClusterStatus(summaryOnly = false) {
     return (dispatch, getState) => {
-        const managerAccessor = new Manager(getState().manager);
+        const { app, manager, pages } = getState();
+        const managerAccessor = new Manager(manager);
+        const fetchOnlySummary = summaryOnly && !isClusterStatusWidgetOnPage(app.currentPageId, pages);
         dispatch(requestClusterStatus());
         return managerAccessor
-            .doGet('/cluster-status')
+            .doGet(`/cluster-status${fetchOnlySummary ? '?summary=true' : ''}`)
```
Comment (kubama) [inline app/actions/clusterStatus.js]: I'm not sure if it would work equally, but maybe it could be simplified to:
``.doGet(`/cluster-status?summary=${fetchOnlySummary}`)``

**Comment by qooban** [comment]
@kubama Thanks for review and comments, I see the point in most of them. As for now I merged the PR in current shape, but I'm going to reply to your comments and implement requested changes after weekend, as well as update cluster status system tests.

File: app/actions/clusterStatus.js (hunk 1 above)
Comment (qooban) [reply app/actions/clusterStatus.js]: Yes, `pageId` is string. Thanks for keeping me informed by that shortcuts in lodash.
Optimized according to suggestions.

File: app/actions/clusterStatus.js (hunk 3 above)
Comment (qooban) [reply app/actions/clusterStatus.js]: Can be. Changed.

File: app/actions/clusterStatus.js (hunk 2 above)
Comment (qooban) [reply app/actions/clusterStatus.js]: Changed.

File: app/reducers/clusterStatusReducer.js (hunk 4 above)
Comment (qooban) [reply app/reducers/clusterStatusReducer.js]: I originally thought of such solution, don't remember why I changed it.
Fixed.

File: package.json (hunk 6 above)
Comment (qooban) [reply package.json]: Yes, I had a problem with tests. After short investigation I found that the reason was in Babel's old version. I am aware of the risk.

File: app/actions/clusterStatus.js (hunk 7 above)
Comment (qooban) [reply app/actions/clusterStatus.js]: Fixed.

File: app/actions/clusterStatus.js (hunk 8 above)
Comment (qooban) [reply app/actions/clusterStatus.js]: Yes, you're right, it turned out that it works equally.

File: app/utils/StatusPoller.js (hunk 5 above)
Comment (qooban) [reply app/utils/StatusPoller.js]: Yes, I was aware of that, but didn't want to refactor whole file. Will fix it in new PR.
